File: Backtracking/Permutations/Permutations.go
package main

import (
	"fmt"
)

func permute(nums []int) [][]int {
	var result [][]int
	backtrack(&result, []int{}, nums)
	return result
}

func backtrack(result *[][]int, tempList []int, nums []int) {
	if len(tempList) == len(nums) {
		copyList := make([]int, len(tempList))
		copy(copyList, tempList)
		*result = append(*result, copyList)
		return
	}
	for i := 0; i < len(nums); i++ {
		if contains(tempList, nums[i]) {
			continue
		}
		tempList = append(tempList, nums[i])
		backtrack(result, tempList, nums)
		tempList = tempList[:len(tempList)-1]
	}
}

func contains(slice []int, num int) bool {
	for _, v := range slice {
		if v == num {
			return true
		}
	}
	return false
}

func main() {
	var t int
	fmt.Scan(&t)
	for i := 0; i < t; i++ {
		var n int
		fmt.Scan(&n)
		nums := make([]int, n)
		for j := 0; j < n; j++ {
			fmt.Scan(&nums[j])
		}
		result := permute(nums)
		for _, nextList := range result {
			for _, x := range nextList {
				fmt.Print(x, " ")
			}
			fmt.Println()
		}
	}
}

File: Backtracking/Permutations/Permutations.cpp
#include <algorithm>
#include <iostream>
#include <vector>
using namespace std;

bool contains(const vector<int>& slice, int num) {
  return find(slice.begin(), slice.end(), num) != slice.end();
}

void backtrack(vector<vector<int>>* result, vector<int> tempList,
               const vector<int>& nums) {
  if (tempList.size() == nums.size()) {
    result->push_back(tempList);
    return;
  }
  for (int i = 0; i < nums.size(); ++i) {
    if (contains(tempList, nums[i])) {
      continue;
    }
    tempList.push_back(nums[i]);
    backtrack(result, tempList, nums);
    tempList.pop_back();
  }
}

vector<vector<int>> permute(const vector<int>& nums) {
  vector<vector<int>> result;
  backtrack(&result, {}, nums);
  return result;
}

int main() {
  int t;
  cin >> t;

  for (int i = 0; i < t; ++i) {
    int n;
    cin >> n;
    vector<int> nums(n);
    for (int j = 0; j < n; ++j) {
      cin >> nums[j];
    }
    vector<vector<int>> result = permute(nums);
    for (const auto& nextList : result) {
      for (const auto& x : nextList) {
        cout << x << " ";
      }
      cout << endl;
    }
  }

  return 0;
}
